Replace expression parser

Krzmbrzl opened the pull request:
This PR will replace the currently existing RegEx-based expression parser with a "proper" parser that is defined by means of an EBNF-like grammar using [Boost Spirit X3](https://www.boost.org/doc/libs/develop/libs/spirit/doc/x3/html/index.html).

This will make the parser implementation much easier to adapt in the future and also a lot easier to understand (one only has to look at the grammar in order to understand the format of valid input.

Fixes #145

bimalgaudel commented on SeQuant/core/expr.hpp:
Do we really need all these constructors? Just curious.

```
Product(ExprPtrList factors, Flatten flatten_tag = Flatten::Yes)
explicit Product(Range &&rng, Flatten flatten_tag = Flatten::Yes)
Product(T scalar, ExprPtrList factors, Flatten flatten_tag = Flatten::Yes)
Product(Iterator begin, Iterator end, Flatten flatten_tag = Flatten::Yes)
Product(T scalar, Iterator begin, Iterator end, Flatten flatten_tag = Flatten::Yes)
```

Krzmbrzl replied:
I doubt that we really need overloads for initializer lists, but I for my part needed the range-constructor, but with the ability to give a scalar. And since the scalar was chosen to be the first argument in the interface, it can't be defaulted, without also defaulting everything else (which of course could also be done - provided one accepts obtaining a default-constructible `Product` class).
However, I didn't want to mess with the current interface design and thus decided to simply add a new constructor.

Krzmbrzl commented on .github/workflows/cmake.yml:
```diff
@@ -10,9 +10,9 @@ jobs:
       matrix:
         build_type : [ Release, Debug ]
-        os : [ macos-latest, ubuntu-20.04 ]
+        os : [ macos-latest, ubuntu-22.04 ]
```
We can also stick with 20.04 and install Boost externally instead of via the Ubuntu repos. Upgrading the CI just seemed to be the simplest solution.

bimalgaudel replied:
What's up the Ubuntu repo's Boost version in 20.04?

Krzmbrzl replied:
It's too old (1.71, but we need 1.72). See https://packages.ubuntu.com/focal/libboost-dev

bimalgaudel replied:
The parser you wrote requires at least 1.72?

Krzmbrzl replied:
Yes - that's why I increased the minimum Boost version to 1.72. Or more specifically: prior Boost versions contain a bug (https://github.com/boostorg/spirit/issues/533) that makes the grammar fail to compile and this bug was fixed in Boost 1.72

bimalgaudel replied:
I suggest we hard-code the fix of that issue in SeQuant whenever the Boost version is <1.72.
The issue is fixed [in this thread](https://github.com/boostorg/spirit/issues/533#issuecomment-530197276).

Setting the minimum Boost version to 1.72 in SeQuant also requires doing the same in mpqc4 (and any other programs that use SeQuant).

@Krzmbrzl

Krzmbrzl replied:
That seems very error prone and not future-proof. Also I don't see why the other lobs would also have to increase their minimum Boost version? If SeQuant is used somewhere, the minimum version will be dictated by it and otherwise the newer versions are not need :thinking:

bimalgaudel commented on tests/unit/test_parse_expr.cpp:
```diff
@@ -112,8 +87,8 @@ TEST_CASE("TEST_PARSE_EXPR", "[parse_expr]") {
     REQUIRE(parse_expr(L"-t{i1;a1}")->is<Product>());
     REQUIRE(*expr == *parse_expr(L"t{\ti1, \ti2; \na1,\t a2 \t}"));
-    REQUIRE_NOTHROW(parse_expr(L"t⁔1{a1;i1}"));
-    REQUIRE_NOTHROW(parse_expr(L"t¹{a1;i1}"));
```
I referred to [this page](https://en.wikipedia.org/wiki/Unicode_subscripts_and_superscripts), while implementing the parser in RegEx. I could not find a more suitable character [class specification either](https://en.cppreference.com/w/cpp/regex/regex_traits/lookup_classname). So I supported any utf8 character to be a valid label character as long as it is not a punctuation, space or a control character. Also the first character of a multi-character label could not be a digit.

Krzmbrzl replied:
Nvm - just found out I can also use character ranges. That'll do nicely.

Krzmbrzl commented on external/versions.cmake:
```diff
@@ -3,8 +3,8 @@
-set(SEQUANT_TRACKED_BOOST_VERSION 1.68)
-set(SEQUANT_TRACKED_BOOST_PREVIOUS_VERSION 1.68)
+set(SEQUANT_TRACKED_BOOST_VERSION 1.72)
+set(SEQUANT_TRACKED_BOOST_PREVIOUS_VERSION 1.72)
```
Not sure whether this might have to stay at 1.68? What's the workflow with this previous-version thingy?

Krzmbrzl commented:
@bimalgaudel I think I have addressed all your review comments (as far as they were directed towards me) and I also merged with current master and extended the parser to deal with things recently introduced (conjugated variables, arrows, sub- and superscript chars in names)

evaleev commented on SeQuant/core/parse/parse.cpp:
```diff
@@ -0,0 +1,219 @@
+//
+
+#include "SeQuant/core/parse_expr.hpp"
+#include "ast.hpp"
```
nonconforming #include, generic names like `ast.hpp` are likely to produce clashes ... please convert to `SeQuant/core/parse/ast.hpp`, here and elsewhere

Krzmbrzl replied:
Already done in https://github.com/ValeevGroup/SeQuant/pull/142

Also, strictly speaking this is guaranteed to never include the wrong file as includes in quotes are always first searched in the current directory. It's only when it is not found there that it looks elsewhere in the include directories for it, at which point issues with name clashes can arise.

evaleev replied:
> Also, strictly speaking this is guaranteed to never include the wrong file as includes in quotes are always first searched in the current directory. It's only when it is not found there that it looks elsewhere in the include directories for it, at which point issues with name clashes can arise.

The standard does not guarantee this: https://en.cppreference.com/w/cpp/preprocessor/include ...  but I might be overly paranoid

Krzmbrzl replied:
Hm. I think I've read it somewhere, but I never actually checked the standard myself. Anyway, better safe than sorry :)
